[c10d] Bring down PG NCCL on failure in abortTimedOutCollectives

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64242
* #64241

It appears that this is only run in async error handling mode, so
there is no need to wait for the work cleanup thread to run and handle this
error if we know the work object encountered an error. We might as well bring
the process down in this case which is what the work cleanup loop would do
anyways. The purpose of doing it here is because occasionally other collectives
may run before the work cleanup loop does, resulting in confusing NCCL error
messages/crashes when we actually just want to bring down the process.

Differential Revision: [D30659941](https://our.internmc.facebook.com/intern/diff/D30659941/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

rohan-varma committed:
[c10d] Bring down PG NCCL on failure in abortTimedOutCollectives

It appears that this is only run in async error handling mode, so
there is no need to wait for the work cleanup thread to run and handle this
error if we know the work object encountered an error. We might as well bring
the process down in this case which is what the work cleanup loop would do
anyways. The purpose of doing it here is because occasionally other collectives
may run before the work cleanup loop does, resulting in confusing NCCL error
messages/crashes when we actually just want to bring down the process.

Differential Revision: [D30659941](https://our.internmc.facebook.com/intern/diff/D30659941/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -577,6 +577,11 @@ void ProcessGroupNCCL::abortTimedOutCollectives(
+      // Handle Exceptions on failed GPU operations by bringing the process
+      // down.
+      if (!terminateProcessGroup_.load()) {
+        work.handleNCCLGuard();
```
This will not terminate the process since the exception will just be caught here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L585 and be logged.

Also, why do we need to check for `terminateProcessGroup_.load()`? One more thing, this can be done before all the abort logic right? Basically if the work is timed out we should just bring the process down.

Btw, don't we need to check for `asyncErrorHandling_` being true before we do this?

rohan-varma replied:
> This will not terminate the process since the exception will just be caught here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L585 and be logged.

Good point, should we try to work around this by changing the handling there to let this exception through and crash? We can probably case it on some const string and check the error message for that string, and if it matches, crash the process instead of just logging it. 

> Btw, don't we need to check for asyncErrorHandling_ being true before we do this?

`abortTimedOutCollectives` is only called if `asyncErrorHandling_`, see https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L645

rohan-varma replied:
> Also, why do we need to check for terminateProcessGroup_.load()? One more thing, this can be done before all the abort logic right? Basically if the work is timed out we should just bring the process down.

The current call to handleNCCLGuard in the workCleanupThread also checks for process group termination, I think the idea is that if concurrently the process group destructor has run and the PG is being destroyeed, there is no real need to bring down the process. cc @osalpekar for this understanding. 

Regarding still aborting NCCL communicators, shouldn't we still abort them as a "best practice" instead of leaving them unaborted? I guess it doesn't make too much of a difference though.

rohan-varma committed:
Update on "[c10d] Bring down PG NCCL on failure in abortTimedOutCollectives"


It appears that this is only run in async error handling mode, so
there is no need to wait for the work cleanup thread to run and handle this
error if we know the work object encountered an error. We might as well bring
the process down in this case which is what the work cleanup loop would do
anyways. The purpose of doing it here is because occasionally other collectives
may run before the work cleanup loop does, resulting in confusing NCCL error
messages/crashes when we actually just want to bring down the process.

Differential Revision: [D30659941](https://our.internmc.facebook.com/intern/diff/D30659941/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Bring down PG NCCL on failure in abortTimedOutCollectives"


It appears that this is only run in async error handling mode, so
there is no need to wait for the work cleanup thread to run and handle this
error if we know the work object encountered an error. We might as well bring
the process down in this case which is what the work cleanup loop would do
anyways. The purpose of doing it here is because occasionally other collectives
may run before the work cleanup loop does, resulting in confusing NCCL error
messages/crashes when we actually just want to bring down the process.

Differential Revision: [D30659941](https://our.internmc.facebook.com/intern/diff/D30659941/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Bring down PG NCCL on failure in abortTimedOutCollectives"


It appears that this is only run in async error handling mode, so
there is no need to wait for the work cleanup thread to run and handle this
error if we know the work object encountered an error. We might as well bring
the process down in this case which is what the work cleanup loop would do
anyways. The purpose of doing it here is because occasionally other collectives
may run before the work cleanup loop does, resulting in confusing NCCL error
messages/crashes when we actually just want to bring down the process.

Differential Revision: [D30659941](https://our.internmc.facebook.com/intern/diff/D30659941/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Bring down PG NCCL on failure in abortTimedOutCollectives"


It appears that this is only run in async error handling mode, so
there is no need to wait for the work cleanup thread to run and handle this
error if we know the work object encountered an error. We might as well bring
the process down in this case which is what the work cleanup loop would do
anyways. The purpose of doing it here is because occasionally other collectives
may run before the work cleanup loop does, resulting in confusing NCCL error
messages/crashes when we actually just want to bring down the process.

Differential Revision: [D30659941](https://our.internmc.facebook.com/intern/diff/D30659941/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

[ghstack-poisoned]

rohan-varma commented:
cc @pritamdamania87

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -594,9 +602,16 @@ void ProcessGroupNCCL::ncclCommWatchdog() {
+              << err;
+    // Errors containing handleNCCLGuardLogStr indicate that the
+    // process is unhealthy and needs to be taken down.
+    if (strstr(err, handleNCCLGuardLogStr) != nullptr) {
```
This is a bit hacky, can we just use `std::exit` on line 592 instead after logging the reason why we are taking the process down?

rohan-varma replied:
We could do this, but won't it introduce inconsistency between the 2 watchdogs in how they bring down the process for nccl_async_error_handling? I am not sure if it is preferable to log + std::exit or crash with an exception, but we are doing the latter currently.

pritamdamania87 replied:
If we want to do this via an exception, let us define a custom exception type for this and throw that. Then here we can have multiple catch statements, one for the custom exception type in which case we rethrow and one for `std::exception` where we perform the rest of the handling.

rohan-varma commented:
cc @pritamdamania87
